src/dataflow/render/mod.rs

```diff
@@ -983,528 +1006,16 @@ where
-                    .chain(next_vals.iter().map(|i| prior_arities[*input] + *i))
-                    .collect();
-
-                // We exploit the demand information to restrict `prev` to its demanded columns.
-                let (prev_keyed, es) = joined.map_fallible({
-                    move |row| {
-                        let datums = row.unpack();
-                        let temp_storage = RowArena::new();
-                        let key = Row::try_pack(
-                            prev_keys.iter().map(|e| e.eval(&datums, &temp_storage)),
-                        )?;
-                        let row = Row::pack(prev_vals.iter().map(|i| datums[*i]));
-                        Ok((key, row))
-                    }
-                });
-                errs = errs.concat(&es);
-                let prev_keyed = prev_keyed
-                    .arrange_named::<OrdValSpine<_, _, _, _>>(&format!("JoinStage: {}", input));
-
-                match self.arrangement(&inputs[*input], &next_keys[..]) {
-                    Some(ArrangementFlavor::Local(oks, es)) => {
-                        joined = prev_keyed.join_core(&oks, move |_keys, old, new| {
-                            let prev_datums = old.unpack();
-                            let next_datums = new.unpack();
-                            // TODO: We could in principle apply some predicates here, and avoid
-                            // constructing output rows that will be filtered out soon.
-                            Some(Row::pack(
-                                prev_datums
-                                    .iter()
-                                    .chain(next_vals.iter().map(|i| &next_datums[*i])),
-                            ))
-                        });
-                        errs = errs.concat(&es.as_collection(|k, _v| k.clone()));
-                    }
-                    Some(ArrangementFlavor::Trace(_gid, oks, es)) => {
-                        joined = prev_keyed.join_core(&oks, move |_keys, old, new| {
-                            let prev_datums = old.unpack();
-                            let next_datums = new.unpack();
-                            // TODO: We could in principle apply some predicates here, and avoid
-                            // constructing output rows that will be filtered out soon.
-                            Some(Row::pack(
-                                prev_datums
-                                    .iter()
-                                    .chain(next_vals.iter().map(|i| &next_datums[*i])),
-                            ))
-                        });
-                        errs = errs.concat(&es.as_collection(|k, _v| k.clone()));
-                    }
-                    None => {
-                        panic!("Arrangement alarmingly absent!");
+                RelationExpr::ArrangeBy { input, keys } => {
+                    if keys
+                        .iter()
```

Comment:
Earlier we were calling `ensure_rendered` on the input of the `keys` vector was empty; why is this no longer necessary?

Reply:
Good question! Probably a bug. Tbh, it's not clear what an arrangement with no keys means (the list contains the requested arrangements), though that should probably be documented one way or the other. Will make sure it maintains the same behavior for now.